algorithms/bug1.py

```diff
@@ -0,0 +1,338 @@
+    curr_y: Numeric
+        Current y location of the point.
+    
+    Returns
+    -------
+    True: Boolean
+        Robot completed the loop around the obstacle.
+    False: Boolean
+        Robot did not loop around the obstacle yet.
+    '''
+    points = []
+    if (curr_x+1 < max_index and grid[curr_x+1][curr_y] != -1):
+        points.append((curr_x+1, curr_y))
+    if (curr_x+1 < max_index and curr_y-1 > -1 and grid[curr_x+1][curr_y-1] != -1):
+        points.append((curr_x+1, curr_y-1))
+    if (curr_y-1 > -1 and grid[curr_x][curr_y-1] != -1):
+        points.append((curr_x, curr_y-1))
+    if (curr_x-1 > -1 and curr_y-1 > -1 and grid[curr_x-1][curr_y-1] != -1):
+        points.append((curr_x-1, curr_y-1))
+    if (curr_x-1 > -1 and grid[curr_x-1][curr_y] != -1):
+        points.append((curr_x-1, curr_y))
+    if (curr_x-1 > -1 and curr_y+1 < max_index and grid[curr_x-1][curr_y+1] != -1):
+        points.append((curr_x-1, curr_y+1))
+    if (curr_y+1 < max_index and grid[curr_x][curr_y+1] != -1):
+        points.append((curr_x, curr_y+1))
+    if (curr_x+1 < max_index and curr_y+1 < max_index and grid[curr_x+1][curr_y+1] != -1):
+        points.append((curr_x+1, curr_y+1))
+        
+    if (initial_x, initial_y) in points:
+        return True
+    return False
+    
+
+
+def bug1(grid, m, start, goal):
+    '''
+    Bug 1 algorithm. Robot starts going towards the goal in a straight line. If 
+    it hits an obstacle, it follows the boundary and loops around it and then 
+    goes to the point which is closest to the goal. Then repeats. 
+    
+    Parameter
+    ---------
+    grid: 
+        2D map containing the robot, goal and obstacles.
+    m: 
+        Map object.
+    start:
+        Starting coordinates of the robot. 
+    goal:
+        Coordinates of goal.
+        
+    Returns
+    -------
+    Path taken by the robot.    
+    '''
+    path = np.zeros( (m.pps, m.pps) )
+    goal_x = goal[0]
+    goal_y = goal[1]
+    
+    curr_x = start[0]
+    curr_y = start[1]
+    
+    prev_x = curr_x
+    prev_y = curr_y
+
+    while True:
+        initial_x = curr_x
+        initial_y = curr_y
+        # Goes straight towards the goal until it hits an obstacle. 
+        while(grid[curr_x][curr_y] != -1):
+            initial_x = curr_x
+            initial_y = curr_y
+            # Getting closest point to the goal which is in the Moore 
+            #  Neighborhood of the point. 
+            curr_x, curr_y = next_point(curr_x, curr_y, goal, path)
+            path[initial_x][initial_y] = 20
+            
+            if (reached_goal(grid, curr_x, curr_y)):
+                return True, path
+        
+        curr_x = initial_x
+        curr_y = initial_y
+
+        distances = [((curr_x, curr_y), get_distance(curr_x, curr_y, goal))]
+        initial_reached = False
+        
+        # Follows boundary of the obstacle
+        while(not initial_reached):
+        
+            # Checks right, down, left and right
```

Comment:
Why not diagonals?

Reply:
A robot has 8 neighbors, and when it follows the boundary and checks for valid points, it starts checking from 3rd column second row and goes clockwise. 
-1 = obstacle 
20 = path 
Using the current algorithm wherein it only checks the 4 directions,  the robot will go from (4,3) to (4,4) and then (4,5).

If the diagonals were used then it would go from (4,3) to (3,4) and when it reached (3,4), it's going to check the right cell first, which is (4,4) and seeing that's it's unvisited, it's going to o go there.
When it is at (4,4), it's going to check all it's directions and realize that it has nowhere to go, and hence, it's going to get stuck and go into and infinite loop.